### ISSUE TITLE
Parallel conversions and synchronization

### ISSUE THREAD

**Issue by cdekker** [unoconv/unoconv#225]
Currently I am using Unoconv together with OpenOffice on a production webserver running Ubuntu to generate PDF files from dynamically written ODT documents.

Initially I ran into problems without using Unoconv that OpenOffice would get randomly stuck and unresponsive without completing the conversion.

I had hoped that Unoconv would solve this problem but that seems not to be the case. From time to time OpenOffice instances still seem to hang and operations never complete resulting in unexpected behavior from my web application.

It seems only 1 process can call Unoconv / OpenOffice at the same time _system wide_!

As a temporary workaround I have created a locking mechanism in my (Ruby on Rails) web application using a Mutex-Redis lock.

Is this an issue of Unoconv that needs to be fixed, or am I simply not using Unoconv correctly? If you have any idea on how to properly implement this use case, please let me know.

**Comment by SergeyTsalkov** [comment]
I'm shocked that openoffice won't let you start more than one instance on a machine. I have 256GB of RAM, RAIDed SSDs, dozens of CPU cores, and I'm really limited to doing my file conversions one at a time?

**Comment by danmilon** [comment]
Is that true for `unoconv --listener` too?

**Comment by SergeyTsalkov** [comment]
Yes -- it won't start more than one listener at a time.

**Comment by danmilon** [comment]
I suppose one openoffice instance cannot convert more than one document at a time, which is reasonable. Maybe we can do something like `unoconv --listener --pool X` which has a pool of 5 OpenOffice instances, so it can do `X` conversions simultaneously?

**Comment by SergeyTsalkov** [comment]
I'd settle for ```unoconv --listener``` not dying if there's another listener already running!

**Comment by danmilon** [comment]
Actually for me it works kinda normal. Running two conversions at the same time spawns two openoffice instances that work in parallel.

**Comment by SergeyTsalkov** [comment]
If you run ```unoconv --listener``` in two separate shells, what do you get? Doesn't one of them immediately exit?

**Comment by danmilon** [comment]
Indeed. Previously I was referring to a single listener handling two requests at the same time. When trying to spawn a second listener this happens:

```
$ ./unoconv/unoconv --listener --port 1337 -vvv
Verbosity set to level 3
Using office base path: /usr/lib/libreoffice
Using office binary path: /usr/lib/libreoffice/program
Start listener on 127.0.0.1:1337
Existing LibreOffice listener found, nothing to do.
```

I suppose it should not be annoyed by existing listeners.

**Comment by cdekker** [comment]
For me, the problem is not the launching of multiple listeners.
I only launch 1 listener during boot on my server.

The problems occur when 2 conversions are sent off to the single listener simultaneous. OpenOffice chokes *sometimes* and either fails or hangs silently.

I would love if unoconv had some kind of option to maintain a local queue with options to prioritize and or block or process asynchronously. But I understand if this is beyond the scope of the project.

**Comment by lorenzyannick** [comment]
Hi @cdekker 
I detect the same issue on my server recently. When a project need more and more conversion in a short time => sounds like simultaneous bugs you're facing.
Did you find something interesting to prevent it ?
Thx

**Comment by dagwieers** [comment]
I would report this to the LibreOffice project and find out what they have to say about it.
Maybe there is a very simple workaround ?

I'd also like to know why the listener cannot be used safely simultaneously.

**Comment by al3x-edge** [comment]
@dagwieers Was there ever a conversation started with the LibreOffice project?

**Comment by sockmonk** [comment]
FWIW, here's my set of hacks to keep this going:
* I run the conversions in a background job, in a queue with only one worker.
* I have a cron job run every 10 minutes that tries to start unoconv (in case it died).
* I have a cron job run once every hour that kills unoconv (in case it locks up).
Since I don't care how long it takes so long as the conversion happens, this works well enough, and if the queue gets behind, it will eventually catch up. It would be nice to have a more stable solution though.

**Comment by gspringtech** [comment]
Is this a possible answer to the multiprocessing question?
https://ask.libreoffice.org/en/question/7284/multiple-instances-for-c-automation/

**Comment by cdekker** [comment]
The past few months I have been rewriting the entire document generation code to step away from OpenOffice and use Prawn to generate the documents programmatically. Obviously, it is less expressive than OpenOffice, but for my situation it is good enough.

Currently I run both systems simultaneously and am slowly converting all my clients from OpenOffice to the Prawn generator. I can't wait to be able to erradicate horrible OpenOffice from my server.

Unoconv has helped a LOT in battling OpenOffice's shortcomings, but there is only so much it can do. For anyone interested in using OpenOffice in a production-like environment, I have only this to say: rethink and start over now, before it is too late and costing you much more time.

It is an absolute shame that idempotent use (read-only) of a small program requires a system wide lock on the profile directory...

**Comment by darlingbot** [comment]
Indeed - parallel conversion of documents is not available. So if this problem is not solvable (maybe only for a now) - which alternatives are available for converting *.doc, *.docx and *.odt documents into *.pdf?

**Comment by pakizito** [comment]
Hi,
Indeed, we ran into this issues also... now I'm looking into the docker alternative to see if it works... I tried to find anything good in pure java that actually converts doc/docx into pdf and there's nothing... or at least that I know of... if someone has an idea of how to really scale a conversion service I would love to know it... thank you very much.

**Comment by hewu2008** [comment]
https://blog.art-of-coding.eu/starting-multiple-openoffice-instances/

it seems that `-userid `parameter can be used to invoke mutiple openoffice instance

**Comment by hewu2008** [comment]
@pakizito have you tried docker alternative?

**Comment by pengqi** [comment]
@dagwieers  Any progressing?

**Comment by Vampouille** [comment]
> I'm looking into the docker alternative

I try to integrate unoconv in a docker image, it only works with ubuntu:xenial image

**Comment by jaishankarm** [comment]
@dagwieers  Can we achive parallel convertion doc to pdf now ? Any fixed done ?

if yes please explain or provide sample code to achive parallel convertion using libre office

**Comment by ozeranskiy** [comment]
@dagwieers, hi, for me this question is also relevant.

**Comment by WillemWW** [comment]
> Is this a possible answer to the multiprocessing question?
> https://ask.libreoffice.org/en/question/7284/multiple-instances-for-c-automation/

works

**Comment by serverdevil** [comment]
I have figured out how to run multiple unoconv invoked listeners for the same office installation. Taking cue from https://ask.libreoffice.org/en/question/42975/how-can-i-run-multiple-instances-of-sofficebin-at-a-time/ making the following changes allowed me to run multiple listeners and making conversion calls to them in parallel did return successfully.
`"-env:UserInstallation=file:///c:/work/temp/lop%s/" % op.port`

I added this to all `Popen` calls. I'm using the port of listener to have designated user homes for each listener. Note that this does not require any modification to `bootstrap.ini` of the office installation as mentioned in a lot of forum posts out there.

My non-existent python skills hinder in me submitting a pull request for this. Maybe someone can work this out in an environment independent manner and commit the code.

**Comment by NicolasGoeddel** [comment]
Would be nice to see an improvement here. We often convert big writer files to PDF which can take up to 30 minutes conversion time or even more. I can change the Popen calls by myself or even add an optional parameter to the whole thing but as this is an open issue since 2014 it would be more cool to add it to the source on Github, so everyone can benefit from it.

**Comment by NicolasGoeddel** [comment]
In case someone is interested: 

- Add a new default value to the `__init__` method of the `Options` class:

```
self.userPath = None
``` 

- Change the getopt-line to this:

```
            opts, args = getopt.getopt (args, 'c:Dd:e:F:f:hi:Llo:np:s:T:t:u:vV',
                ['connection=', 'debug', 'doctype=', 'export=', 'field=', 'format=',
                 'help', 'import=', 'listener', 'no-launch', 'output=',
                 'outputpath', 'password=', 'pipe=', 'port=', 'preserve',
                 'server=', 'timeout=', 'show', 'stdin', 'stdout', 'template', 'user=',
                 'verbose', 'version'] )
```

- Add a new `elif` condition:

```
            elif opt in ['-u', '--user']:
                self.userPath = arg
```

- Change the code between `try:` and `info(2, '%s listener successfully started....` to this:

```
                product = self.svcmgr.createInstance("com.sun.star.configuration.ConfigurationProvider").createInstanceWithArguments("com.sun.star.configuration.ConfigurationAccess", UnoProps(nodepath="/org.openoffice.Setup/Product"))
                if product.ooName not in ('LibreOffice', 'LOdev') or LooseVersion(product.ooSetupVersion) <= LooseVersion('3.3'):
                    ooprocArgs = [office.binary, "-headless", "-invisible", "-nocrashreport", "-nodefault", "-nofirststartwizard", "-nologo", "-norestore", "-accept=%s" % op.connection]
                else:
                    ooprocArgs = [office.binary, "--headless", "--invisible", "--nocrashreport", "--nodefault", "--nofirststartwizard", "--nologo", "--norestore", "--accept=%s" % op.connection]
                
                if op.userPath is not None :
                    ooprocArgs.append("-env:UserInstallation=%s" % op.userPath)
                ooproc = subprocess.Popen(ooprocArgs, env=os.environ)
```

**Comment by kuraga** [comment]
@NicolasGoeddel , thanks! Works in 16 threads!!!

@unoconv @regebro @cdekker @dagwieers would yo accept a patch?

**Comment by L3o-pold** [comment]
@kuraga could you please create a PR?

**Comment by kuraga** [comment]
@L3o-pold hah, seems like it's already in `master`. I've just used old Ubuntu version. Sorry.

**Comment by kuraga** [comment]
NicolasGoeddel Works! But unfortunately still crash and freezes on several threads...

**Comment by NicolasGoeddel** [comment]
@kuraga I did not merge that into master. Seems like someone else made a pull request out of my suggestion.
We have no freezes anymore on our production setup, even with documents up to 3000 pages. But I know that there are other issues on Libreoffice itself where it runs into an endless loop. I already created issues on their tracker because of that. I guess it has to do with oversized images which can not be layouted headless. I don't know why.

**Comment by NicolasGoeddel** [comment]
That is the utility I use to create a PDF from ODT: https://github.com/NicolasGoeddel/utilities/blob/master/libreoffice.py

But at the moment it still uses the parameter how I defined it in this thread: `-u` or `--user`. You need to change that to `--user-profile` in line 47.

**Comment by kuraga** [comment]
@NicolasGoeddel ,
>  But I know that there are other issues on Libreoffice itself where it runs into an endless loop.

Yes, LibreOffice itself. But not only endless loops: 
* `SIGSEGV`, `exitcode` = 0, exit status = -11;
* `uno.DisposedException: Binary URP bridge already disposed`, `exitcode` = 6, exit status = 1;
* `SystemError during update-indexes phase: Couldn't instantiate python representation of stru
ctured UNO type com.sun.star.lang.DisposedException`, `exitcode` = 1, exit status = 1;
* failed assertions at multithread libraries (e.g. `tpp.c:84: __pthread_tpp_change_priority: Assertion 'new_prio == -1 || (new_prio >= fifo_min_prio && new_prio <= fifo_max_prio)' failed.`), `exitcode` = 0, exit status = -6;
* \<unknown message\>, `exitcode` = \<unknown\>, exit status = 2;
* zombies (if start with `--accept` (socket));
* maybe more.

Reproducible: stable if threads are >1; almost never if =1. (Remember: use different `--user`)

```
$ soffice --version
LibreOffice 6.3.1.2 30(Build:2)
```
From here: https://launchpad.net/~libreoffice/+archive/ubuntu/ppa

> I already created issues on their tracker because of that. 

Could you point links, please?

**Comment by kuraga** [comment]
@NicolasGoeddel , and I really don't understand: how do they interact each other (if user profiles are different)? Single instance, several... What is the difference?

**Comment by NicolasGoeddel** [comment]
@kuraga I reported a bug here which could be solved for us by using unoconv: https://bugs.documentfoundation.org/show_bug.cgi?id=124798
But that does not mean that it will work for everyone. But it helped on our production line.
I never experienced these exit code bugs you mentioned. So I can not help you in this point.

**Comment by kuraga** [comment]
Thanks!

No, that's else. Your issue is caused a concrete file. But my:
> Reproducible: stable if threads are >1; almost never if =1. (Remember: use different `--user`)

**Comment by gulien** [comment]
Hello @NicolasGoeddel 

Works great so far. In my case, there are a lot of `soffice.bin` zombies. Same for you?

**Comment by NicolasGoeddel** [comment]
@gulien We don't have any zombies here. But we only convert ODT to PDF, nothing more. I don't know what you are doing exactly.

**Comment by gulien** [comment]
@NicolasGoeddel this was an issue with Docker and PID 1.

Still, some `soffice.bin` (childs of `unoconv` I guess) are not correctly killed if unoconv is killed 🤔

**Comment by kuraga** [comment]
@gulien ,

> Works great so far.

> this was an issue with Docker

So works or works in Docker? And what exactly does work? Because [it doesn't for me](https://github.com/unoconv/unoconv/issues/225#issuecomment-532864826).

**Comment by gulien** [comment]
@kuraga 

Does work:
- starting multiple instances of `unoconv` using the `--user-profile` flag (and therefore multiple conversions)
- for Docker, you must start the main process (your program) with a utility like https://github.com/krallin/tini to reap the zombie processes (in our case: `soffice.bin`)

Does not work:
- killing a `unoconv` process seems not propagating the signal to its  `soffice.bin` child process

**Comment by kuraga** [comment]
@NicolasGoeddel ,
(if interesting) I've found a criteria for [my issues](https://github.com/unoconv/unoconv/issues/225#issuecomment-532864826).

So, calling `unoconv` in parallel using different `--user-profilie`s. It did have that crashes since I've not pointed `--port`s (so they're the same). Calling with different `--user-profile`s and `--port`s/`--host`&`--port`s/--`--accept`s doesn't have that crashes. What do you think - is this LibreOffice bug?

@gulien you said it works for you.
1. One more time: _without_ Docker?
2. If so, which `--port`s did you point?

Thanks.
